File: packages/webstreamr/lib/webstreamr/source/vidsrc.dart
/// Port of webstreamr/src/source/VidSrc.ts
library;

import '../types.dart';
import '../utils/id.dart';
import '../webstreamr_parse.dart';
import 'source.dart';

class VidSrcSource extends Source {
  VidSrcSource(super.fetcher);

  @override
  String get id => 'vidsrc';
  @override
  String get label => 'VidSrc';
  @override
  int? get useOnlyWithMaxUrlsFound => 0;
  @override
  List<String> get contentTypes => const ['movie', 'series'];
  @override
  List<CountryCode> get countryCodes => const [CountryCode.multi];
  @override
  String get baseUrl => 'https://vidsrc-embed.ru';

  @override
  Future<List<SourceResult>> handleInternal(
      Context ctx, String type, Id id) async {
    return requireRustResolveSource('vidsrc', type, id);
  }
}

File: crates/webstreamr/src/sources/vidsrc.rs
use super::{SourceEmbed, SourceRequest};
use crate::types::MediaType;

/// MBG WebStreamr VidSrc source host (not VSEmbed / vsembed.su).
const BASE_URL: &str = "https://vidsrcme.ru";

fn base_url() -> String {
    utils::provider_runtime::webstreamr_base("vidsrc")
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| BASE_URL.to_string())
}

/// MBG shape: `/embed/movie/{id}` · `/embed/tv/{id}/{season}-{episode}` (IMDb preferred).
pub fn resolve(req: &SourceRequest) -> Vec<SourceEmbed> {
    let base = base_url().trim_end_matches('/').to_string();
    let id = req
        .imdb_id
        .as_deref()
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .or_else(|| req.tmdb_id.map(|t| t.to_string()));
    let Some(id) = id else {
        return Vec::new();
    };

    let url = match &req.media_type {
        MediaType::Movie => format!("{base}/embed/movie/{id}"),
        MediaType::Series => {
            let season = req.season.unwrap_or(1);
            let episode = req.episode.unwrap_or(1);
            format!("{base}/embed/tv/{id}/{season}-{episode}")
        }
    };
    vec![SourceEmbed {
        url,
        title: None,
        country_codes: vec!["multi".into()],
        referer: None,
        priority: None,
        height: None,
        bytes: None,
    }]
}
